src/System.Private.CoreLib/shared/System/Collections/ObjectModel/Collection.cs

```diff
@@ -162,6 +170,61 @@ protected virtual void SetItem(int index, T item)
             items[index] = item;
         }
 
+        protected virtual void InsertItemsRange(int index, IEnumerable<T> collection)
+        {
+            if (collection == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.list);
+            }
+
+            if (items.IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
+            }
+
+            if ((uint)index > (uint)items.Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index, ExceptionResource.ArgumentOutOfRange_ListInsert);
+            }
+
+            int i = 0;
+            foreach (T item in collection)
+            {
+                Insert(index + i, item);
+                i++;
+            }
+        }
+
+        protected virtual void RemoveItemsRange(int index, int count)
+        {
+            if (items.IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
+            }
+
+            if ((uint)index > (uint)items.Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index, ExceptionResource.ArgumentOutOfRange_ListInsert);
+            }
+
+            int length = items.Count;
```

Comment:
Actually, I think we should follow what `List<T>.RemoveRange` does, which is to throw an `ArgumentException`:
https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.removerange?view=netframework-4.7.2
https://github.com/dotnet/coreclr/blob/14caedf474161891c09e740448f07e541e7a95b5/src/System.Private.CoreLib/shared/System/Collections/Generic/List.cs#L927-L928

Reply:
@stephentoub, @safern - what do you think?

Reply:
I would agree with this recommendation. I think we should follow the List exception patterns to keep the API consistent with other similar APIs.

Reply:
@ahsonkhan I haven't started this work yet, should we go ahead and start this change or do we need to wait for input from @stephentoub or @safern

Reply:
Go for it :)

Reply:
To clarify, instead of this:

> If `index + count` > `int.MaxValue` is true, maybe we should constrain to `items.Count`.

We should throw `ArgumentException`. ( `index + count > item.Count` implies `index + count > int.MaxValue` since `item.Count <= int.MaxValue`).

Reply:
@ahsonkhan your reasoning makes a lot of sense and I like going this route, but I need to make one modification because my int.MaxValue test case was failing.

Old:
`if (index + count > int.MaxValue)`

New:
`if ((Int64)index + (Int64)count > int.MaxValue)`

In the case where `index + count > int.MaxValue` it was not properly working since `int` (Int32) does not support an integer value greater than int.MaxValue. By casting them to `Int64` in the case where the combination of `index + count` does equal more than `int.MaxValue` the math calculation completes which allows us to determine the resulting value is or is not greater than `int.MaxValue`

Reply:
You can change this to: `if (index > int.MaxValue - count)`

But, ideally you should compare to `items.Count`, not `int.MaxValue`

`if (index > items.Count - count)`

This is effectively the same test in `List<T>` (see code I linked above).

Reply:
I copied over the logic wrong here as it is different in the code, my mistake. 

The code I updated reads as (latest revision in PR):
`if ((Int64)index + (Int64)count > items.Count)`

Does this work for you?

Reply:
> Does this work for you?

 
The cast to `Int64` is unnecessary, if you re-write it like this:
`if (index > items.Count - count)`